Improve dynamics section of docs (#2271)
qutip/qutip: doc/guide/dynamics/dynamics-monte.rst

will stop either after all errors bars on expectation values are under ``0.01``, 1000
trajectories are computed or 10 minutes have passed, whichever comes first. When a
single values is passed, it is used as the absolute value of the tolerance.
When a pair of values is passed, it is understood as an absolute and relative
tolerance pair. For even finer control, one such pair can be passed for each ``e_ops``.
For example:

.. code-block::

    data = mcsolve(H, psi0, times, c_ops, e_ops=e_ops,  target_tol=[
        (1e-5, 0.1),
        (0, 0),
    ])

will stop when the error bars on the expectation values of the first ``e_ops`` are
under 10% of their average values.

If after computation of some trajectories, it is determined that more are needed, it
is possible to add trajectories to existing result by adding result together:

.. code-block::

    >>> run1 = mcsolve(H, psi, times, c_ops, e_ops=e_ops, ntraj=25)
    >>> print(run1.num_trajectories)
    25
    >>> run2 = mcsolve(H, psi, times, c_ops, e_ops=e_ops, ntraj=25)
    >>> print(run2.num_trajectories)
    25
    >>> merged = run1 + run2
    >>> print(merged.num_trajectories)
    50

> Note that this merging operation only check that the result are compatible:
> same e_ops and same tlist. It does not check that the same initial state or
> Hamiltonian where used.

Reviewer: ```suggestion
Note that this merging operation only checks that the result are compatible --
i.e. that the ``e_ops`` and ``tlist`` are the same. It does not check that the same initial state or
Hamiltonian where used.
```

    plt.show()

The original sampling algorithm samples the no-jump trajectory on average 96.7%
of the time, while the improved sampling algorithm only does so once.

.. _monte-seeds:

Reproducibility
---------------

> For reproducibility of Monte-Carlo computations it is possible to set the seed

Reviewer: ```suggestion
For reproducibility of Monte-Carlo computations it is possible to set the seed of the random
number generator:
```

.. code-block::

    >>> res1 = mcsolve(H, psi0, tlist, c_ops, e_ops=e_ops, seeds=1, ntraj=1)
    >>> res2 = mcsolve(H, psi0, tlist, c_ops, e_ops=e_ops, seeds=1, ntraj=1)
    >>> res3 = mcsolve(H, psi0, tlist, c_ops, e_ops=e_ops, seeds=2, ntraj=1)
    >>> np.allclose(res1, res2)
    True
    >>> np.allclose(res1, res3)
    False

> The ``seeds`` parameter can either be an integer or numpy SeedSequence, which
> will then be used to create seeds for each trajectories. Or a list of
> interger/SeedSequence with one seed for each trajectories. Seeds available in
> the result object can be used to redo the same evolution:

Reviewer: ```suggestion
The ``seeds`` parameter can either be an integer or a numpy ``SeedSequence``, which
will then be used to create seeds for each trajectory. Alternatively it may be a list of
intergers or ``SeedSequence``s with one seed for each trajectories. Seeds available in
the result object can be used to redo the same evolution:
```

.. code-block::

    >>> res1 = mcsolve(H, psi0, tlist, c_ops, e_ops=e_ops, ntraj=10)
    >>> res2 = mcsolve(H, psi0, tlist, c_ops, e_ops=e_ops, seeds=res1.seeds, ntraj=10)
    >>> np.allclose(res1, res2)
    True

.. _monte-parallel:

Running trajectories in parallel
--------------------------------

Monte-Carlo evolutions often need hundreds of trajectories to obtain sufficient
statistics. Since all trajectories are independent of each other, they can be computed
in parallel. The option ``map`` can take ``"serial"``, ``"parallel"`` or ``"loky"``.

> Both ``"parallel"`` and ``"loky"`` compute trajectories on multiple cpus using

Reviewer: ```suggestion
Both ``"parallel"`` and ``"loky"`` compute trajectories on multiple CPUs using
```

respectively the `multiprocessing <https://docs.python.org/3/library/multiprocessing.html>`_
and `loky <https://loky.readthedocs.io/en/stable/index.html>`_ python modules.

.. code-block::

    >>> res_par = mcsolve(H, psi0, tlist, c_ops, e_ops=e_ops, options={"map": "parallel"}, seeds=1)
    >>> res_ser = mcsolve(H, psi0, tlist, c_ops, e_ops=e_ops, options={"map": "serial"}, seeds=1)
    >>> np.allclose(res_par.average_expect, res_ser.average_expect)
    True

> Note the when running in parallel, the order in which the trajectories are added

Reviewer: ```suggestion
Note that when running in parallel, the order in which the trajectories are added
```

    [SeedSequence(entropy=1,spawn_key=(1,),),
     SeedSequence(entropy=1,spawn_key=(0,),),
     SeedSequence(entropy=1,spawn_key=(2,),)]

    >>> print(res_ser.seeds[:3])
    [SeedSequence(entropy=1,spawn_key=(0,),),
     SeedSequence(entropy=1,spawn_key=(1,),),
     SeedSequence(entropy=1,spawn_key=(2,),)]

Photocurrent
------------

The photocurrent, previously computed using the ``photocurrent_sesolve`` and
``photocurrent_sesolve`` functions, are now included in the output of
:func:`.mcsolve` as ``result.photocurrent``.

.. plot::
    :context: close-figs

    times = np.linspace(0.0, 10.0, 200)
    psi0 = tensor(fock(2, 0), fock(10, 8))
    a  = tensor(qeye(2), destroy(10))
    sm = tensor(destroy(2), qeye(10))
    e_ops = [a.dag() * a, sm.dag() * sm]
    H = 2*np.pi*a.dag()*a + 2*np.pi*sm.dag()*sm + 2*np.pi*0.25*(sm*a.dag() + sm.dag()*a)
    data = mcsolve(H, psi0, times, [np.sqrt(0.1) * a], e_ops=e_ops)

    plt.figure()
    plt.plot((times[:-1] + times[1:])/2, data.photocurrent[0])
    plt.title('Monte Carlo Photocurrent')
    plt.xlabel('Time')
    plt.ylabel('Photon detections')
    plt.show()

.. openmcsolve:

Open Systems
------------

> ``mcsolve`` can be used to study system with have measured and dissipative
> interaction with the bath.  This is done by using a liouvillian including the
> dissipative interaction instead of an Hamiltonian.

Reviewer: ```suggestion
``mcsolve`` can be used to study systems which have measurement and dissipative
interactions with their environment.  This is done by passing a Liouvillian including the
dissipative interaction to the solver instead of a Hamiltonian.
```
